### ISSUE TITLE
Images built from docker fail on `image.expand`

### ISSUE THREAD

**Issue by alexhrescale** [apptainer/singularity#1076]
## Version of Singularity:

[2.4-HEAD.g81bf030](https://github.com/singularityware/singularity/tree/2.4)

### Expected behavior

image is resizable like an image created via `image.create`

### Actual behavior

```
ext2fs_open2: Bad magic number in super-block
/sbin/e2fsck: Superblock invalid, trying backup blocks...
/sbin/e2fsck: Bad magic number in super-block while trying to open bar.img
```

#### failure case

```
$ singularity build --writable fail.img docker://centos:latest
Docker image path: index.docker.io/library/centos:latest
Cache folder set to /root/.singularity/docker
Importing: base Singularity environment
Importing: /root/.singularity/docker/sha256:d9aaf4d82f249dc101a6638ff5177fe926cdebfa6c42d874dfa5029533da0e72.tar.gz
Importing: /root/.singularity/metadata/sha256:53ee036c793c94bbd5b7061c45c6fa375522042e624577f232fda1d88778f242.tar.gz
Creating empty Singularity writable container 206MB
Creating empty 257MiB image file: fail.img
Formatting image with ext3 file system
Image is done: fail.img
Building Singularity image...
Singularity container built: fail.img
Cleaning up...
$ singularity image.expand fail.img
Expanding image by 768MB
Checking image's file system
e2fsck 1.42.9 (28-Dec-2013)
ext2fs_open2: Bad magic number in super-block
/sbin/e2fsck: Superblock invalid, trying backup blocks...
/sbin/e2fsck: Bad magic number in super-block while trying to open fail.img

The superblock could not be read or does not describe a correct ext2
filesystem.  If the device is valid and it really contains an ext2
filesystem (and not swap or ufs or something else), then the superblock
is corrupt, and you might try running e2fsck with an alternate superblock:
    e2fsck -b 8193 <device>
```

#### working case

```
$ singularity image.create ok.img
Creating empty 768MiB image file: ok.img
Formatting image with ext3 file system
Image is done: ok.img
$ singularity image.expand ok.img
Expanding image by 768MB
Checking image's file system
e2fsck 1.42.9 (28-Dec-2013)
Pass 1: Checking inodes, blocks, and sizes
Pass 2: Checking directory structure
Pass 3: Checking directory connectivity
Pass 4: Checking reference counts
Pass 5: Checking group summary information
ok.img: 11/49152 files (0.0% non-contiguous), 7387/196608 blocks
Resizing image's file system
resize2fs 1.42.9 (28-Dec-2013)
Resizing the filesystem on ok.img to 393216 (4k) blocks.
The filesystem on ok.img is now 393216 blocks long.

Image is done: ok.img
```

**Comment by GodloveD** [comment]
Hi @alexhrescale.  I've been doing some research on this issue.  It seems that the problem is not specific to containers built from Docker Hub, but manifests with any writable image.  It also seems that Singularity actually _does_ perform the expand operation as requested and the image is still usable afterward.  But it just complains as it does it.  

Can you corroborate these findings?

**Comment by jekriske-lilly** [comment]
```
vagrant@vagrant:~$ sudo singularity build --writable fail.img docker://centos:latest
Docker image path: index.docker.io/library/centos:latest
Cache folder set to /root/.singularity/docker
[1/1] |===================================| 100.0%
Importing: base Singularity environment
Importing: /root/.singularity/docker/sha256:d9aaf4d82f249dc101a6638ff5177fe926cdebfa6c42d874dfa5029533da0e72.tar.gz
Importing: /root/.singularity/metadata/sha256:53ee036c793c94bbd5b7061c45c6fa375522042e624577f232fda1d88778f242.tar.gz
Creating empty Singularity writable container 208MB
Creating empty 260MiB image file: fail.img
Formatting image with ext3 file system
Image is done: fail.img
Building Singularity image...
Singularity container built: fail.img
Cleaning up...
vagrant@vagrant:~$ sudo singularity exec fail.img df -h
Filesystem                    Size  Used Avail Use% Mounted on
OverlayFS                     497M     0  497M   0% /
/dev/mapper/vagrant--vg-root   38G  2.2G   34G   7% /tmp
udev                          477M     0  477M   0% /dev
tmpfs                         497M     0  497M   0% /dev/shm
tmpfs                         100M  3.2M   97M   4% /etc/resolv.conf
vagrant@vagrant:~$ sudo singularity image.expand fail.img
Expanding image by 768MB
Checking image's file system
e2fsck 1.42.13 (17-May-2015)
ext2fs_open2: Bad magic number in super-block
/sbin/e2fsck: Superblock invalid, trying backup blocks...
/sbin/e2fsck: Bad magic number in super-block while trying to open fail.img

The superblock could not be read or does not describe a valid ext2/ext3/ext4
filesystem.  If the device is valid and it really contains an ext2/ext3/ext4
filesystem (and not swap or ufs or something else), then the superblock
is corrupt, and you might try running e2fsck with an alternate superblock:
    e2fsck -b 8193 <device>
 or
    e2fsck -b 32768 <device>
vagrant@vagrant:~$ sudo singularity exec fail.img df -h
Filesystem                    Size  Used Avail Use% Mounted on
OverlayFS                     497M     0  497M   0% /
/dev/mapper/vagrant--vg-root   38G  3.0G   33G   9% /tmp
udev                          477M     0  477M   0% /dev
tmpfs                         497M     0  497M   0% /dev/shm
tmpfs                         100M  4.5M   95M   5% /etc/resolv.conf
```

The filesystem isn't expanded even though the file is. `expand` under 2.3 would have done this correctly and without complaining.

```
vagrant@vagrant:~$ sudo /tmp/sing/bin/singularity pull --size 500 docker://centos:latest
Initializing Singularity image subsystem
Opening image file: centos-latest.img
Creating 500MiB image
Binding image to loop
Creating file system within image
Image is done: centos-latest.img
Docker image path: index.docker.io/library/centos:latest
Cache folder set to /root/.singularity/docker
[1/1] |===================================| 100.0%
Importing: base Singularity environment
Importing: /root/.singularity/docker/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4.tar.gz
Importing: /root/.singularity/docker/sha256:d9aaf4d82f249dc101a6638ff5177fe926cdebfa6c42d874dfa5029533da0e72.tar.gz
Importing: /root/.singularity/metadata/sha256:2062d56428495ab9ee869306e0f6549267fb5870c68f3255269155b77b2f202f.tar.gz
Done. Container is at: centos-latest.img
vagrant@vagrant:~$ sudo /tmp/sing/bin/singularity exec centos-latest.img df -h
Filesystem                    Size  Used Avail Use% Mounted on
OverlayFS                     1.0M     0  1.0M   0% /
/dev/mapper/vagrant--vg-root   38G  2.2G   34G   7% /tmp
udev                          477M     0  477M   0% /dev
tmpfs                         497M     0  497M   0% /dev/shm
tmpfs                          16M     0   16M   0% /tmp/sing/var/singularity/mnt/session
/dev/loop0                    477M  197M  255M  44% /tmp/sing/var/singularity/mnt/container
tmpfs                         1.0M     0  1.0M   0% /tmp/sing/var/singularity/mnt/overlay
tmpfs                         100M  4.5M   95M   5% /etc/resolv.conf
vagrant@vagrant:~$ sudo /tmp/sing/bin/singularity expand centos-latest.img
Initializing Singularity image subsystem
Opening image file: centos-latest.img
Expanding image by 768MiB
Binding image to loop
Checking file system
e2fsck 1.42.13 (17-May-2015)
Pass 1: Checking inodes, blocks, and sizes
Pass 2: Checking directory structure
Pass 3: Checking directory connectivity
Pass 4: Checking reference counts
Pass 5: Checking group summary information
/dev/loop0: 8402/128016 files (21.6% non-contiguous), 225944/512000 blocks
Resizing file system
resize2fs 1.42.13 (17-May-2015)
Resizing the filesystem on /dev/loop0 to 1298432 (1k) blocks.
The filesystem on /dev/loop0 is now 1298432 (1k) blocks long.

Image is done: centos-latest.img
vagrant@vagrant:~$ sudo /tmp/sing/bin/singularity exec centos-latest.img df -h
Filesystem                    Size  Used Avail Use% Mounted on
OverlayFS                     1.0M     0  1.0M   0% /
/dev/mapper/vagrant--vg-root   38G  3.0G   33G   9% /tmp
udev                          477M     0  477M   0% /dev
tmpfs                         497M     0  497M   0% /dev/shm
tmpfs                          16M     0   16M   0% /tmp/sing/var/singularity/mnt/session
/dev/loop0                    1.2G  198M  960M  18% /tmp/sing/var/singularity/mnt/container
tmpfs                         1.0M     0  1.0M   0% /tmp/sing/var/singularity/mnt/overlay
tmpfs                         100M  4.5M   95M   5% /etc/resolv.conf
```

**Comment by GodloveD** [comment]
Ah.  I see.  Thank you for clarification.  When you do a `pull` you are just downloading whatever format the image happens to be in.  You are not converting image types.  So this totally makes sense.  Basically anytime we are working with an ext3 image created using 2.4 expand fails. 

I'll have a look and see if I can figure out what is causing this bug.  Thank you.

**Comment by jekriske-lilly** [comment]
Keeping in mind  a build issued by 2.4 can be expanded using 2.3 expand

**Comment by GodloveD** [comment]
OK.  I checked on this and I think it is because version 2.3 used to mount the file system and then expand it in a mounted state.  2.4 just works directly on the image.  I'm going to talk to @gmkurtzer about this and figure out the best way to proceed.  

But in the meantime, have you tried using the `--sandbox` option?  This will build your container into a directory so that you don't have to worry about expanding an image when it is going to run out of room.  

I'm pretty sure that we're discouraging the use of ext3 moving forward and encouraging folks to use `--sandbox` instead.  Do you think that would work for you?

**Comment by jekriske-lilly** [comment]
I can't speak for the OP, but I've been using `--sandbox` myself to get around this. It works for me. 

I was more worried about the inconsistent behavior between native images and those pulled from Docker. I'm looking into singularity for our HPC environment. This and #764 (proxy) stood out as potential issues if I point scientists to the singularity docs. The examples won't always work out of the box.

**Comment by alexhrescale** [comment]
Confirming `--sandbox` works, it's magnificent!

**Comment by moskalenko** [comment]
I got the same error with a ubuntu based ext3 container with sg-2.4.0, will have to try sandbox.

**Comment by alexhrescale** [comment]
I have been using `--sandbox` exclusively since learning about it. I can't really think of a reason to not use sandbox at this point.

**Comment by jekriske-lilly** [comment]
If it's not officially supported, it'd be good to better detect and handle the issue.

**Comment by moskalenko** [comment]
It doesn't always work. Right now sandbox Ubuntu 16.04 Containers are failing any apt actions on our cluster. The ext3 containers are fine.

**Comment by dasizeman** [comment]
Just ran into this same issue... it seems like it comes from the fact that image.expand is expecting an ext2 filesystem instead of ext3.  If the `--writable` ext3 images are basically broken then they this should really be removed from the documentation, right?  It seems like ext3 writable images might be a dead feature currently.  

/cc @vsoch @gmkurtzer can either of you comment on this?

**Comment by gmkurtzer** [comment]
Apologies for the latent reply. Supercomputing has many of us way behind on emails. I am also CC'ing @GodloveD to see if he can throw in his $0.02 here, but I do want to reiterate that the EXT3 based images are not a dead feature, but there is a bug here.

Thanks!

**Comment by GodloveD** [comment]
@dasizeman I agree this should be removed from documentation.  The only thing that `image.expand` is good for anymore is expanding an image that has not been built (i.e. has been created but has not had a header added).  This may be useful if you are using the image as an overlay.  I've talked to @gmkurtzer about this and we need to adjust the help docs and add some errors to reflect this new usage.

**Comment by cinek810** [comment]
I meet this issue today and looks like it was quite easy to solve. In my case the issue was 31 bytes offset preventing e2fsck/resize2fs from recognizing that it's ext filesystem. I was able to resize it doing mount bind with proper offset. I wrote a [blog post](https://funinit.wordpress.com/2018/10/01/solved-singularity-2-6-fails-to-resize-writable-container/) post on how to debug it and eventually patch for 2.6. 

I'm just not sure on how to automatically determine the ext3 offset..

**Comment by jmstover** [comment]
@cinek810 ... For 2.x I believe it should always be 32bytes.  For example run:

```
$ head -c 32 centos7.img
#!/bin/bin/env run-singularity
```

The initial part of the image file is setup so you can run an image as a executable. So, for example ... 

`./centos7 -a 1 -b 2 -c 3`

Is the equivalent to:  `singularity run centos7 -a 1 -b 2 -c 3`

**Comment by dtrudg** [comment]
Believed fixed in 2.6 via #2150
